pkg/rbac/parser.go

```diff
@@ -44,8 +47,54 @@ type Rule struct {
 	URLs      []string `marker:"urls,optional"`
 }
 
+// ruleKey represents the resources and non-resources a Rule applies.
+type ruleKey string
+
+// key normalizes the Rule and returns a ruleKey object.
+func (r *Rule) key() ruleKey {
+	r.normalize()
+	s := []string{
+		strings.Join(r.Groups, "&"),
+		strings.Join(r.Resources, "&"),
+		strings.Join(r.URLs, "&"),
+	}
+	return ruleKey(strings.Join(s, "+"))
+}
+
+// addVerbs adds new verbs into a Rule.
+// The duplicates in `r.Verbs` will be removed, and then `r.Verbs` will be sorted.
+func (r *Rule) addVerbs(verbs []string) {
+	r.Verbs = removeDupAndSort(append(r.Verbs, verbs...))
+}
+
+// normalize removes duplicates from each field of a Rule, and sorts each field.
+func (r *Rule) normalize() {
+	r.Groups = removeDupAndSort(r.Groups)
+	r.Resources = removeDupAndSort(r.Resources)
+	r.Verbs = removeDupAndSort(r.Verbs)
+	r.URLs = removeDupAndSort(r.URLs)
+}
+
+// removeDupAndSort removes duplicates in s, sorts the items, and returns a
+// new slice of strings.
+func removeDupAndSort(s []string) []string {
+	set := make(map[string]bool)
+	for _, str := range s {
```

Comment:
we try to avoid single-letter variables for anything but method receivers, especially in cases where you've got a lot of similar types.

Reply:
Fixed it! Thanks!